### PR TITLE
WIP: basic wrapper of gesvd

### PR DESCRIPTION
**Pull request by JerryChen97**
Currently not able to run samples.jl, due to the lack of wrappers for parts of essential utility routines.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,4 @@
 .DS_Store
 /Manifest.toml
+binary/
```
Comment (Roger-luo) [inline .gitignore]: We need to put the binary in `build` folder, just simply rename the `binary` -> `build`, so we could use BB (BinaryProvider) for MAGMA later.

Comment (JerryChen97) [reply .gitignore]: I changed the name just now.

**Comment by Roger-luo** [comment]
which part is runable now? can you separate the implementation and test into `src` and `test` folder, and make a working example ASAP.

please read the documentation and check how to write tests. And then you could run the test by

```
] test MAGMA
```

File: samples/samples.jl [hunk 2]
```diff
@@ -0,0 +1,213 @@
+# low-level wrappers of the MAGMA library
+include("common.jl")
+
+
+# magma_init
+function magmaInit()
+	ccall((:magma_init, libmagma),Cint,())
```
Comment (Roger-luo) [inline samples/samples.jl]: So after you move these implementation to `src`, we need to have some unit tests that make sure each function works or not. For working tests you can use `@test` to mark them
for broken tests use `@test_broken` for temporary.

Comment (JerryChen97) [reply samples/samples.jl]: OK, I will immediately go check some testing examples in the official Julia documentation.

Comment (JerryChen97) [reply samples/samples.jl]: Do you have any clue where and how the GPU buffer error could ever come?
It's so weird. I don't think I have met with it one month ago...
BTW yesterday I opened a new branch in my repository for debugging this problem.
I checked all the results on Google but no one situation really matches with ours...
Maybe later I should ask on slack

ERROR: LoadError: ArgumentError: cannot take the CPU address of a GPU buffer
Stacktrace:
 [1] convert(::Type{Ptr{Float32}}, ::CUDAdrv.Mem.DeviceBuffer) at /home/jerry/.julia/packages/CUDAdrv/WVU1H/src/memory.jl:83
 [2] unsafe_convert(::Type{Ptr{Float32}}, ::CUDAdrv.Mem.DeviceBuffer) at /home/jerry/.julia/packages/CUDAdrv/WVU1H/src/memory.jl:32
 [3] unsafe_convert(::Type{PtrOrCuPtr{Float32}}, ::CUDAdrv.Mem.DeviceBuffer) at /home/jerry/.julia/packages/CUDAdrv/WVU1H/src/pointer.jl:115
 [4] magmaSgesvd(::Int64, ::Int64, ::Int64, ::Int64, ::CuArray{Float32,2}, ::Int64, ::CuArray{Float32,2}, ::CuArray{Float32,2}, ::Int64, ::CuArray{Float32,2}, ::Int64, ::CuArray{Float32,1}, ::Int64, ::CuArray{Float32,1}) at /home/jerry/MAGMA.jl-1/samples/MAGMA.jl/src/MAGMA.jl:49
 [5] gesvd!(::Int64, ::Int64, ::CuArray{Float32,2}, ::Int64, ::Int64, ::Int64) at /home/jerry/MAGMA.jl-1/samples/MAGMA.jl/src/MAGMA.jl:174
 [6] top-level scope at none:0
in expression starting at /home/jerry/MAGMA.jl-1/samples/test/test.jl:30

**Comment by Roger-luo** [comment]
Don't use space in your path. And make sure you follow the workflow in docs instead of using push path to LOAD PATH

**Comment by Roger-luo** [comment]
Let's make each PR small first. We only need a working svd here. And then add other features in another PR

**Comment by JerryChen97** [comment]
> 
> 
> Don't use space in your path. And make sure you follow the workflow in docs instead of using push path to LOAD PATH
> 
> You need to use the Pkg which is the package manager and use dev or julia --project to load the module in local path. Do not try to load a single file it is not the package.

I will try using Pkg to load it as a package tomorrow.
I edited the LOAD_PATH just for temporary expediency since at that time there were so many different problems emerging along with the very first version of testing that I totally forgot to study how to utilize Pkg...

**Comment by JerryChen97** [comment]
> 
> 
> Let's make each PR small first. We only need a working svd here. And then add other features in another PR
> 
> You could just use dense (lower case) for dense linalg. It is more consistent with what we have in stdlib.

Oh I got it. Indeed I should break it down into different PRs.

BTW should we always keep the same hierarchical structure as people did in the standard library? The reason why I design our src like this is that I hope to separate the fundamental functions from each other. Days ago, I felt great pain when checking the source codes in JuliaGPU to locate some certain inconspicuous methods and their wrappers, which are mixed together in a few Julia scripts... Maybe if we directly use the directory structure to indicate the logical relationship and keep every single Julia file small but precise, it will be much more friendly especially for those who are not so familiar with the MAGMA source codes?
Anyway the above is just my personal opinion and I don't even know whether it complies with our general standards or not... If it will definitely bring much inconvenience or even potential trouble, I'd better change it now.

**Comment by Roger-luo** [comment]
You feel pain mainly because you are using the wrong workflow. Use Revise, edit/which macro for inspecting code and use module instead of load path.

**Comment by Roger-luo** [comment]
And I don't think we need to care about whoever aren't familiar with magma source code in our source, only experts would read the source code.

**Comment by Roger-luo** [comment]
The main problem in your case is incorrect path name, this may cause serious problem on many platforms. please don't use space in the path anymore.

**Comment by Roger-luo** [comment]
And we follow the convention of using lower case path (this matters for platforms like Mac OS), and it really doesn't matter whether to put them in a single file or multiple directories, but the point is, your convention is very rare in software. It might cause issues in the future, I just want to get rid of these potential problem by simply use the right coding style.

**Comment by JerryChen97** [comment]
> 
> 
> And we follow the convention of using lower case path (this matters for platforms like Mac OS), and it really doesn't matter whether to put them in a single file or multiple directories, but the point is, your convention is very rare in software. It might cause issues in the future, I just want to get rid of these potential problem by simply use the right coding style.

Oh I think it's quite a problem... I will fix it👌

**Comment by Roger-luo** [comment]
It generally looks fine to me. But I don't think it is necessary to have a very long name like `dense_linear_algebra` the whole point of this package is linear algebra, and we don't want to repeat it in the code again and again (the DRY principal). But I'll merge this first, so you could keep working on other things.

**Pull request merged**